src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -132,13 +132,23 @@ class RelocateEmbeddedPointers : public BitMapClosure {
     size_t field_offset = size_t(bit_offset - _start_idx) * sizeof(address);
     address* ptr_loc = (address*)(_buffered_obj + field_offset);
 
-    address old_p = *ptr_loc;
+    address old_p_with_tags = *ptr_loc;
+    address old_p = MetaspaceClosure::strip_tags(old_p_with_tags);
+    uintx tags = MetaspaceClosure::decode_tags(old_p_with_tags);
     address new_p = _builder->get_buffered_addr(old_p);
+    bool implicitly_nulled = false;
+    if (new_p != nullptr) {
+      new_p = MetaspaceClosure::add_tags(new_p, tags);
+    } else if (old_p != nullptr) {
+      implicitly_nulled = true;
+    }
 
-    log_trace(cds)("Ref: [" PTR_FORMAT "] -> " PTR_FORMAT " => " PTR_FORMAT,
-                   p2i(ptr_loc), p2i(old_p), p2i(new_p));
+    log_trace(cds)("Ref: [" PTR_FORMAT "] -> " PTR_FORMAT " => " PTR_FORMAT " %zu%s",
+                   p2i(ptr_loc), p2i(old_p) + tags, p2i(new_p), tags,
+                   implicitly_nulled ?  ", ref to excluded objects is implicitly nulled" : "");
```

Comment:
Hi Igor, could you take https://github.com/openjdk/leyden/commit/334471da3ea3c0071ae305c8c1ff2be6dd34fca6 as well? That cleans up the log message.

Reply:
Yup. Done! Thank you!